### PR TITLE
Add USPTO documentation

### PR DESCRIPTION
**Pull request by AggelosMargkas**
Add my first example of results of data that run through the whole USPTO data.

Add the sql query into the uspto examples file. This SQL query is designed to extract information about the applicant population by country and year, focusing on the top 5 countries in terms of patent counts for the year 2022.

Add the graph representing the evolution of the population and show it on the README.

### CODE REVIEW

**Review by dspinellis — CHANGES_REQUESTED** [review]
Nice! I like the clear SQL code. Please also add the Python script that generates the chart.

**Comment by AggelosMargkas** [comment]
> Nice! I like the clear SQL code. Please also add the Python script that generates the chart.

Added that, should I add the csv data file as well?

**Comment by dspinellis** [comment]
Great! The CSV data is a generated file. We add the chart (which is also generated) so that it can appear in the README file. I'll ensure that the example can run automatically as part of integration testing and get back to you, if needed.

File: docs/app-eg.rst [hunk 1]
```diff
@@ -61,3 +61,5 @@ The following examples are available.
    works <https://github.com/dspinellis/alexandria3k/tree/main/examples/tawfik>`__
 -  `Fully-indexed Crossref database
    sample <https://github.com/dspinellis/alexandria3k/tree/main/examples/sample>`__
+-  `Top 5 USPTO applicants population by country and year
```
Comment (dspinellis) [inline docs/app-eg.rst:64]: Top five US Patent Office applications by …

File: docs/cli-eg.rst [hunk 2]
```diff
@@ -63,6 +63,54 @@ where ``count-year-type.sql`` contains:
    SELECT type AS name, Sum(number) FROM counts
      GROUP BY type
 
+Count USPTO top 5 applicant countries from 2005 to 2022
```
Comment (dspinellis) [inline docs/cli-eg.rst:66]: Obtain top-five US Patent Office application countries

File: docs/cli-eg.rst [hunk 3]
```diff
@@ -98,6 +146,36 @@ you can re-seed the random number generator for each sample with
...
 similar.
 
+For the USPTO dataset sampling can be achieved by providing a 
+tuple as its argument where the first value is a designator
+string that can be either "path" or "container". This way
+sampling can be done either for compressed files which represents 
+the weekly issued patents (path) or for the patents(container) which
+are located inside the issued file. The second value of the tuple can
+be an USPTO Zip file path or a container respectively.
+
```
Comment (dspinellis) [inline docs/cli-eg.rst:156]: For the USPTO dataset sampling is performed through a 
provided tuple argument named **data**. The tuple's first value is
a designator string that will be either "path" or "container".
Thus, sampling can be done either on compressed files represented
through the weekly issued patents (path) or on the patents' full text
(container) located within the issued file.
The tuple's second value is a USPTO Zip file path or the patent's text, respectively.

File: docs/cli-eg.rst [hunk 4]
```diff
@@ -98,6 +146,36 @@ you can re-seed the random number generator for each sample with
...
+.. note::
+   How to access the 
+   sampling tuple? It is already defined as a variable inside *alexandria3k* and
+   is named **data**. So, the first value of the tuple will data[0].
```
Comment (dspinellis) [inline docs/cli-eg.rst:160]: I covered it above, please remove the note.

File: docs/downloading.rst [hunk 5]
```diff
@@ -41,6 +41,44 @@ download the *summary file* of the ORCID Public Data File
 works on the compressed file; there is no need to expand it (it expands
 to about 0.5 TB).
 
+USPTO data
+~~~~~~~~~~
+
+You can populate a database with the bibliographic text (front page) of each
+patent application published weekly from 2003 to present. This data are published
```
Comment (dspinellis) [inline docs/downloading.rst:48]: These data

File: docs/downloading.rst [hunk 6]
```diff
@@ -41,6 +41,44 @@ download the *summary file* of the ORCID Public Data File
...
+from the US Patent and Trademark Office (USPTO) and hence the data source acronym inside *alexandria3k*.
```
Comment (dspinellis) [inline docs/downloading.rst:49]: inside → used in

File: docs/downloading.rst [hunk 7]
```diff
@@ -41,6 +41,44 @@ download the *summary file* of the ORCID Public Data File
...
+.. code-block:: text
+
+   /data
+   ├── 2023
+   │   ├── ipgb20230131_wk05.zip
+   │   └── ipgb20230207_wk06.zip
+   ├── 2010
+       ├── ipgb20100223_wk08.zip
+       └── ipgb20100302_wk09.zip
+
```
Comment (dspinellis) [inline docs/downloading.rst:80]: Add:
```
The shell script *examples/uspto/fetch.sh* can be used to download all USPTO data.
```

File: examples/uspto/README.md [hunk 8]
```diff
@@ -0,0 +1,5 @@
+# Applicants Population by Year and Country
+
+The graph illustrates the calculated population of applicants who are listed in the records of the US Patent and Trademark Office, focusing specifically on the top five countries of the year 2022.
```
Comment (dspinellis) [inline examples/uspto/README.md:3]: The chart illustrates the number of applicants listed in the records of the US Patent and Trademark Office,
tracking those from the top five countries of the year 2022.

File: examples/uspto/README.md [hunk 9]
```diff
@@ -0,0 +1,5 @@
...
+
+![Applicants Population by Country and year for the Top 5 Countries of 2022](./top5applicantcountries.png)
```
Comment (dspinellis) [inline examples/uspto/README.md:5]: Number of Applicants by Country and Year for the Top-Five Countries of 2022Top-Five

File: examples/uspto/plot_applicants_data.py [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+plt.title("Applicants by Year and Country")
+plt.legend(title="Country", bbox_to_anchor=(1.05, 1), loc="upper left")
+
+# Set the x-axis ticks to display each year
+plt.xticks(pivot_data.columns)
+
+# Show the plot
+plt.grid(True)
+plt.tight_layout()
+plt.show()
```
Comment (dspinellis) [inline examples/uspto/plot_applicants_data.py:46]: Shouldn't the last line save the data into a file?
Can the file be SVG, rather than PNG?

**Review by dspinellis — CHANGES_REQUESTED** [review]
So nice to see this contribution! I suggested a few changes.

**Comment by AggelosMargkas** [comment]
> So nice to see this contribution! I suggested a few changes.

Hello👋 and apologies for my late reply. I updated the changes and I believe they are ready to be seen. I changed the image format accordingly. 

I have this error on the CI log for the installation of the rdbunit.

```
Run git clone --depth=1 https://github.com/dspinellis/rdbunit.git
Cloning into 'rdbunit'...
make: *** No rule to make target 'install'.  Stop.
Error: Process completed with exit code 2.
```

Any idea how to resolve this?

**Pull request merged**